IonHashDotnet/IonHashWriter.cs

```diff
@@ -0,0 +1,342 @@
+
+        public void WriteTimestamp(Timestamp value)
+        {
+            this.HashScalar(IonType.Timestamp, value);
+            this.writer.WriteTimestamp(value);
+        }
+
+        public void WriteValue(IIonReader reader)
+        {
+            this.InternalWriteValue(reader);
+        }
+
+        public void WriteValues(IIonReader reader)
+        {
+            this.InternalWriteValues(reader);
+        }
+
+        private void InternalWriteValue(IIonReader reader, int depth = 0)
+        {
+            IonType type = reader.CurrentType;
+            if (type == IonType.None)
+            {
+                return;
+            }
+
+            if (depth > 0)
+            {
+                string fieldName = reader.CurrentFieldName;
+                if (fieldName != null)
+                {
+                    this.SetFieldName(fieldName);
+                }
+            }
+
+            foreach (var annotation in reader.GetTypeAnnotations())
+            {
+                this.AddTypeAnnotationSymbol(annotation);
+            }
+
+            if (reader.CurrentIsNull)
+            {
+                this.WriteNull(type);
+            }
+            else
+            {
+                switch (type)
+                {
+                    case IonType.Bool:
+                        this.WriteBool(reader.BoolValue());
+                        break;
+                    case IonType.Int:
+                        this.WriteInt(reader.BigIntegerValue());
+                        break;
+                    case IonType.Float:
+                        this.WriteFloat(reader.DoubleValue());
+                        break;
+                    case IonType.Decimal:
+                        this.WriteDecimal(reader.DecimalValue());
+                        break;
+                    case IonType.Timestamp:
+                        this.WriteTimestamp(reader.TimestampValue());
+                        break;
+                    case IonType.Symbol:
+                        this.WriteSymbol(reader.StringValue());
+                        break;
+                    case IonType.String:
+                        this.WriteString(reader.StringValue());
+                        break;
+                    case IonType.Clob:
+                        this.WriteClob(reader.NewByteArray());
+                        break;
+                    case IonType.Blob:
+                        this.WriteBlob(reader.NewByteArray());
+                        break;
+                    case IonType.List:
+                        this.StepIn(IonType.List);
+                        break;
+                    case IonType.Sexp:
+                        this.StepIn(IonType.Sexp);
+                        break;
+                    case IonType.Struct:
+                        this.StepIn(IonType.Struct);
+                        break;
+                    default:
+                        throw new IonHashException("Unexpected type '" + type + "'");
```

Comment:
and `InvalidOperationException` here, too

Reply:
Done.